### ISSUE TITLE
Extractor does not work with the 2019c version of the standard

### ISSUE THREAD

**Issue by fedorov** [innolitics/dicom-standard#11]
I am getting the following error:
<pre>
$ make
cat standard/part03.html | sed -e 's/&nbps;/ /g' -e 's/​//g' > tmp/part03.html
PYTHONPATH=.. python3 extract_ciod_module_data.py tmp/part03.html > tmp/raw_ciod_module_tables.json
PYTHONPATH=.. python3 process_ciods.py tmp/raw_ciod_module_tables.json > dist/ciods.json
PYTHONPATH=.. python3 extract_modules_with_attributes.py tmp/part03.html > tmp/raw_module_attribute_tables.json
PYTHONPATH=.. python3 extract_macros.py tmp/part03.html > tmp/raw_macro_tables.json
PYTHONPATH=.. python3 preprocess_modules_with_attributes.py tmp/raw_module_attribute_tables.json tmp/raw_macro_tables.json > tmp/preprocessed_modules_attributes.json
Traceback (most recent call last):
  File "preprocess_modules_with_attributes.py", line 55, in <module>
    tables_with_macros = expand_all_macros(module_attr_tables, macro_tables)
  File "preprocess_modules_with_attributes.py", line 18, in expand_all_macros
    for table in module_attr_tables]
  File "preprocess_modules_with_attributes.py", line 18, in <listcomp>
    for table in module_attr_tables]
  File "/Users/fedorov/github/dicom-standard/dicom_standard/macro_utils.py", line 22, in expand_macro_rows
    for attr in table['attributes']]
  File "/Users/fedorov/github/dicom-standard/dicom_standard/macro_utils.py", line 22, in <listcomp>
    for attr in table['attributes']]
  File "/Users/fedorov/github/dicom-standard/dicom_standard/macro_utils.py", line 30, in get_attributes_to_insert
    new_attributes = get_macro_attributes(attribute, macros, table_id)
  File "/Users/fedorov/github/dicom-standard/dicom_standard/macro_utils.py", line 53, in get_macro_attributes
    return expand_macro_rows(get_macros_by_id(macro_id, macros, hierarchy_marker), macros)
  File "/Users/fedorov/github/dicom-standard/dicom_standard/macro_utils.py", line 72, in get_macros_by_id
    macro['attributes'] = update_attribute_hierarchy_markers(macro['attributes'], hierarchy_marker)
  File "/Users/fedorov/github/dicom-standard/dicom_standard/macro_utils.py", line 77, in update_attribute_hierarchy_markers
    return [add_marker_to_attr(attribute, marker) for attribute in attributes]
  File "/Users/fedorov/github/dicom-standard/dicom_standard/macro_utils.py", line 77, in <listcomp>
    return [add_marker_to_attr(attribute, marker) for attribute in attributes]
  File "/Users/fedorov/github/dicom-standard/dicom_standard/macro_utils.py", line 81, in add_marker_to_attr
    parsed_attribute_name = BeautifulSoup(attribute['name'], 'html.parser').find('td')
KeyError: 'name'
make: *** [tmp/preprocessed_modules_attributes.json] Error 1
</pre>

**Comment by johndgiese** [comment]
Andrey, thank you for creating an issue for this.  I suspect (without having looked into the error) that it is due to subtle HTML cases which were introduced in the DICOM standard since we last ran our tool.  Occasionally we need to make updates to handle issues like this.

We will try to look into fixing this sometime soon.

**Comment by fedorov** [comment]
Could you comment on why you parse HTML and not the XML representation of the standard? I would think that one should be "cleaner" and less susceptible to "subtle HTML cases", or was your experience different?

**Comment by johndgiese** [comment]
We wanted to be able to extract URLs to particular pages of the HTML version of the standard.  This is used in our "View in Standard" links throughout dicom.innolitics.com.

I believe the XML and HTML are both programmatically generated, so I suspect the differences in parsing them are not too great.

David Clunie was also surprised that we did not use the XML version, for what it is worth; I think it may not be too difficult to generate URLs to the HTML version programmatically.  Note, although we don't currently use them, the tables in the standard often have URL fragments which would let use make the URLs go to the correct location on the page, but we don't currently do this.

**Comment by fedorov** [comment]
Thank you for your explanation! I cc @dclunie here, since I mentioned your tool, and he was wondering about the motivation. I didn't investigate myself, so cannot comment about your approach, but maybe David can consider this as a feedback to help make the future versions of the standard more useable for parsing.

**Comment by fedorov** [comment]
I think I understand your concern about the URLs better now. Understandably, you want to have URLs pointing to the sections of chunked HTML, but I don't think chunking is maintained in the XML version of the standard.

**Comment by dclunie** [comment]
The source of the standard is XML DocBook, and all the other forms (HTML, chunked HTML, PDF, Word, ODT) are programmatically generated from that using the DocBook XSL-T Stylesheets (+/- other post-processors, e.g., to make PDF from XSL-FO).

Paragraph level identifiers for links are specified in the DocBook (xml:id attribute of DocBook para elements) and these get translated to HTML anchor element id attributes in the HTML and chunked HTML.

But the chunking (and hence the relative URL) is generated by the DocBook stylesheets, and not specified declaratively in the XML DocBook. See:

http://www.sagehill.net/docbookxsl/Chunking.html
http://www.sagehill.net/docbookxsl/ChunkingCustomization.html

When generating the DICOM standard output, currently the only chunking customization is:

<xsl:param name="chunk.section.depth" select="4"></xsl:param>

since the default is 3, which makes PS3.3 sections too large.

In short, if you wanted to create your content linked to URLs in the current standard but using the XML DocBook as the source, the URLs of the generated chunked HTML should be predictable, both at the section and paragraph level.

David

**Comment by fedorov** [comment]
@johndgiese @yujanshrestha I started from scratch with some code that parses DocBook XML. It is just a start, and it does not do everything, but I wanted to make you aware of that repo: https://github.com/ImagingDataCommons/dcm-docbook-tools

Comments (and contributions!) are welcomed!

**Comment by johndgiese** [comment]
@fedorov thank you for the update! I'm sorry we have been slow to move forward addressing the bugs with our system.

**Comment by ReeceStevens** [comment]
@fedorov @dclunie thanks for your patience as we investigated the issues with the standard parser!

I have isolated the problem to two changes in the DICOM standard XML source:

1. A new macro table, [Multi-energy CT Processing Attributes][misnamed-macro], was named slightly differently from prior convention. While all other macro tables are named "\<macro name\> Macro Attributes", this table is simply named "\<macro name\> Attributes".  This makes it difficult to distinguish between macro and module attribute tables. To work around this, we have explicitly marked this table as a macro table (see 38effad).

2. Additional fields added to the [Multi-energy CT X-Ray Detector Macro][th-macro] were entered as `th` elements in the table, rather than `td`. We worked around this by parsing both `th` and `td` cells inside table body rows (see d43e146).

These changes are currently in review at #12 and will be merged soon.

[th-macro]: http://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.8.2.2.2.html#table_C.8.2.2-3

[misnamed-macro]: http://dicom.nema.org/medical/dicom/current/output/html/part03.html#sect_C.8.15.3.13

**Comment by dclunie** [comment]
Thanks Reece ... I will correct these errors in the next release of the DICOM standard (2020a) ... David

**Comment by johndgiese** [comment]
I was able to verify that the scripts run to completion. I believe this issue is now addressed! Nice work @ReeceStevens !
